internal/messagequeue/mqpinger.go

```diff
@@ -0,0 +1,243 @@
+package messagequeue
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+)
+
+const (
+	// dontHaveTimeout is used to simulate a DONT_HAVE when communicating with
+	// a peer whose Bitswap client doesn't support the DONT_HAVE response.
+	// If the peer doesn't respond to a want-block within the timeout, the
+	// local node assumes that the peer doesn't have the block.
+	dontHaveTimeout = 5 * time.Second
+
+	// maxExpectedWantProcessTime is the maximum amount of time we expect a
+	// peer takes to process a want and initiate sending a response to us
+	maxExpectedWantProcessTime = 200 * time.Millisecond
+
+	// latencyMultiplier is multiplied by the average ping time to
+	// get an upper bound on how long we expect to wait for a peer's response
+	// to arrive
+	latencyMultiplier = 2
+
+	// maxPingErrorCount is the maximum number of times we can receive an
+	// error while pinging before we give up
+	maxPingErrorCount = 3
+
+	// pingInterval is the time between pings
+	pingInterval = 60 * time.Second
+)
+
+// PeerConnection is a connection to a peer that can be pinged, and the
+// average latency measured
+type PeerConnection interface {
+	// Ping the peer
+	Ping(context.Context) ping.Result
+	// The average latency of all pings
+	Latency() time.Duration
+}
+
+// Stoppable can be stopped
+type Stoppable interface {
+	// Stop the Stoppable - the return value is false if it was already stopped.
+	Stop() bool
+}
+
+// mqPinger pings the peer to measure latency. It uses the latency to set a
+// reasonable timeout for AfterTimeout().
+type mqPinger struct {
```

Comment:
This feels really complicated for what it does. What if we just used a lock instead of dancing with channels?

Reply:
Yeah, let's:

1. Just use a mutex.
2. Append CID/timestamps to a "pending" queue.
3. Have a goroutine that pops elements off this slice from left to right, sleeping until `pending[0].timestamp+timeout` before handling each element.
4. Drop all cancellation logic. We'll work through this queue at a steady rate anyways and we can just re-check if the want is pending when we get to the CID in the queue.

Reply:
Be careful about a want/cancel/want race. If the user requests a CID, cancels the request, then re-requests the CID, we need to make sure we use the _new_ timeout.

That might be a reason to keep some cancellation logic...